src/EVM/SMT.hs

```diff
@@ -961,216 +874,91 @@ getBufs getVal names = foldM getBuf mempty names
-          pure solverInstance
-
--- | Cleanly shutdown a running solver instnace
-stopSolver :: SolverInstance -> IO ()
-stopSolver (SolverInstance _ stdin stdout stderr process) = cleanupProcess (Just stdin, Just stdout, Just stderr, process)
-
--- | Sends a list of commands to the solver. Returns the first error, if there was one.
-sendScript :: SolverInstance -> SMT2 -> IO (Either Text ())
-sendScript solver (SMT2 cmds _) = do
-  sendLine' solver (T.unlines $ fmap toLazyText cmds)
-  pure $ Right()
-
--- | Sends a single command to the solver, returns the first available line from the output buffer
-sendCommand :: SolverInstance -> Text -> IO Text
-sendCommand inst cmd = do
-  -- trim leading whitespace
-  let cmd' = T.dropWhile isSpace cmd
-  case T.unpack cmd' of
-    "" -> pure "success"      -- ignore blank lines
-    ';' : _ -> pure "success" -- ignore comments
-    _ -> sendLine inst cmd'
-
--- | Sends a string to the solver and appends a newline, returns the first available line from the output buffer
-sendLine :: SolverInstance -> Text -> IO Text
-sendLine (SolverInstance _ stdin stdout _ _) cmd = do
-  T.hPutStr stdin (T.append cmd "\n")
-  hFlush stdin
-  T.hGetLine stdout
-
--- | Sends a string to the solver and appends a newline, doesn't return stdout
-sendLine' :: SolverInstance -> Text -> IO ()
-sendLine' (SolverInstance _ stdin _ _ _) cmd = do
-  T.hPutStr stdin (T.append cmd "\n")
-  hFlush stdin
-
--- | Returns a string representation of the model for the requested variable
-getValue :: SolverInstance -> Text -> IO Text
-getValue (SolverInstance _ stdin stdout _ _) var = do
-  T.hPutStr stdin (T.append (T.append "(get-value (" var) "))\n")
-  hFlush stdin
-  fmap (T.unlines . reverse) (readSExpr stdout)
-
--- | Reads lines from h until we have a balanced sexpr
-readSExpr :: Handle -> IO [Text]
-readSExpr h = go 0 0 []
-  where
-    go 0 0 _ = do
-      line <- T.hGetLine h
-      let ls = T.length $ T.filter (== '(') line
-          rs = T.length $ T.filter (== ')') line
-      if ls == rs
-         then pure [line]
-         else go ls rs [line]
-    go ls rs prev = do
-      line <- T.hGetLine h
-      let ls' = T.length $ T.filter (== '(') line
-          rs' = T.length $ T.filter (== ')') line
-      if (ls + ls') == (rs + rs')
-         then pure $ line : prev
-         else go (ls + ls') (rs + rs') (line : prev)
 
+    addElem :: W256 -> W256 -> Map W256 (Map W256 W256) -> (W256 -> W256 -> W256) -> Map W256 (Map W256 W256)
+    addElem addr slot store fun =
+      case Map.lookup addr store of
+        Just m -> Map.insert addr (Map.insert slot (fun addr slot) m) store
+        Nothing -> Map.insert addr (Map.singleton slot (fun addr slot)) store
 
 
--- ** Helpers ** ---------------------------------------------------------------------------------
+    queryValue :: Expr EWord -> IO W256
+    queryValue (Lit w) = pure w
+    queryValue w = do
+      let expr = toLazyText $ exprToSMT w
```

Comment:
I would like to move the Cex datatypes and functions to their own file `Cex.hs`  but because of this call to `exprToSMT`, the Cex parsing depends on SMT encoding and the SMT encoding depends on Cex because the`SMT2` datatype uses the `CexVars` datatype. There are possible workarounds:

1) Keep the `CexVars` datatype in the SMT.hs file. Import SMT from Cex. 
2) Leave the file as is 
3) Encode the value of storage reads when constructing the CexVars records and pass it along with their Expr representation.
4) Pass a callback to `exprToSMT` to the `getStore` function

Thoughts?

Reply:
1 seems fine to me